### PR TITLE
CLI Global Options - Part 1 - bulk commands

### PR DESCRIPTION
**Pull request by tonylee80**
Adds optional global options for `bulk` commands. The next step is to follow up with PRs to refactor the remaining CLI commands to support global options.

```shell
npx medplum bulk export -e Group/all --base-url https://sandbox.bcda.cms.gov --fhir-url-path api/v2/ --token-url https://sandbox.bcda.cms.gov/auth/token --client-id {{clientId}} --client-secret {{clientSecret}}

Patient_data_55881_6ec2a601_7178_4785_ba36_db1e6f40684f_ndjson.ndjson is created
Coverage_data_55881_e798f821_0e15_466a_8cd6_39715a0e93ff_ndjson.ndjson is created
ExplanationOfBenefit_data_55881_8e7ff717_b0bb_4d99_abac_4ca57a861542_ndjson.ndjson is created
``` 

### Summary
🧪🛠️🌐

This pull request refactors the `bulk` command of the `cli` package to use a custom `MedplumCommand` class and a `MedplumClient` instance that can be configured with global options and environment variables. It also adds a new utility function `createMedplumClient` to create and mock the `MedplumClient` instance, and a new property `accessToken` to the `MedplumClientOptions` interface. It updates the test cases and the import statements accordingly.

> _We create the Medplum client with global options_
> _We mock the fetch and the console for our tests_
> _We use the Medplum command to simplify our code_
> _We handle the unauthenticated case with a message of doom_

### Walkthrough
*  Added a new utility function `createMedplumClient` to create a `MedplumClient` instance with global options and environment variables ([link](https://github.com/medplum/medplum/pull/2197/files?diff=unified&w=0#diff-80e8d5a02c540100cc38f9c1240d73742010ee4b2f195e615943ac89d67d5bc7R1-R36), [link](https://github.com/medplum/medplum/pull/2197/files?diff=unified&w=0#diff-d5acb19f7f25aef929d46e05e464d658e4d47f9d1aa5e71b88f76d4f21fa8574R1-R88))
*  Added a new utility function `onUnauthenticated` to handle the case when the `MedplumClient` instance is unauthenticated and print a message to the console ([link](https://github.com/medplum/medplum/pull/2197/files?diff=unified&w=0#diff-80e8d5a02c540100cc38f9c1240d73742010ee4b2f195e615943ac89d67d5bc7R1-R36), [link](https://github.com/medplum/medplum/pull/2197/files?diff=unified&w=0#diff-2fb448d11702324cc59d657d5945d59cd325540eb87b71f2e66a5bc01ada834cR11), [link](https://github.com/medplum/medplum/pull/2197/files?diff=unified&w=0#diff-2fb448d11702324cc59d657d5945d59cd325540eb87b71f2e66a5bc01ada834cL86-L89))
*  Added a new utility module `command.ts` that provides a custom `MedplumCommand` class that inherits from `Command` and adds global options for the `MedplumClient` configuration, and a function `createMedplumCommand` to create a `MedplumCommand` instance with a given name ([link](https://github.com/medplum/medplum/pull/2197/files?diff=unified&w=0#diff-f7784b5d582e6d550a96c1d187f2040c05e8b1928e0f0c0d2fbfcea528af05ceR1-R20))
*  Modified the `bulk` command to use the new utility functions and classes, and to accept options from the command line or from the environment variables ([link](https://github.com/medplum/medplum/pull/2197/files?diff=unified&w=0#diff-aff1e5d2e9becbf0e02038967adcd4314c1c6588ecca9d825de98d5110a42cafL1-R10), [link](https://github.com/medplum/medplum/pull/2197/files?diff=unified&w=0#diff-aff1e5d2e9becbf0e02038967adcd4314c1c6588ecca9d825de98d5110a42cafL22-R25), [link](https://github.com/medplum/medplum/pull/2197/files?diff=unified&w=0#diff-aff1e5d2e9becbf0e02038967adcd4314c1c6588ecca9d825de98d5110a42cafL46-R54), [link](https://github.com/medplum/medplum/pull/2197/files?diff=unified&w=0#diff-aff1e5d2e9becbf0e02038967adcd4314c1c6588ecca9d825de98d5110a42cafL67-R80))
*  Modified the `bulk.test.ts` file to mock the `./util/client` module and the `createMedplumClient` function, and to use the same mocked `MedplumClient` instance for testing the `bulk export` and `bulk import` commands ([link](https://github.com/medplum/medplum/pull/2197/files?diff=unified&w=0#diff-ba0eec9b49430d3db287c8fc6a316004d00df70f9396382fef7d8a1c513824eeL1-R3), [link](https://github.com/medplum/medplum/pull/2197/files?diff=unified&w=0#diff-ba0eec9b49430d3db287c8fc6a316004d00df70f9396382fef7d8a1c513824eeR10), [link](https://github.com/medplum/medplum/pull/2197/files?diff=unified&w=0#diff-ba0eec9b49430d3db287c8fc6a316004d00df70f9396382fef7d8a1c513824eeR133-R134), [link](https://github.com/medplum/medplum/pull/2197/files?diff=unified&w=0#diff-ba0eec9b49430d3db287c8fc6a316004d00df70f9396382fef7d8a1c513824eeL183-R191), [link](https://github.com/medplum/medplum/pull/2197/files?diff=unified&w=0#diff-ba0eec9b49430d3db287c8fc6a316004d00df70f9396382fef7d8a1c513824eeL211))
*  Added a new property `accessToken` to the interface `MedplumClientOptions` in `packages/core/src/client.ts` to allow passing an OAuth Access Token as an option to the `MedplumClient` constructor ([link](https://github.com/medplum/medplum/pull/2197/files?diff=unified&w=0#diff-858339789c798051d9c13620bffce8ce860e7ad8ba18dba1c8711eb1f7d51a55R115-R128))

### CODE REVIEW

File: packages/cli/src/index.ts [hunk 1]
```diff
@@ -84,6 +84,6 @@ if (require.main === module) {
   run().catch((err) => console.error('Unhandled error:', err));
 }
 
-function onUnauthenticated(): void {
+export function onUnauthenticated(): void {
```
Comment (tonylee80) [inline packages/cli/src/index.ts:87]: this should be moved to `utils`

File: packages/cli/src/util/client.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  const clientSecret = options.clientSecret || process.env['MEDPLUM_CLIENT_SECRET'];
+
+  if (clientId && clientSecret) {
+    medplumClient.setBasicAuth(clientId, clientSecret);
+    await medplumClient.startClientLogin(clientId, clientSecret);
+  }
+  return medplumClient;
+}
+
+export function onUnauthenticated(): void {
```
Comment (tonylee80) [inline packages/cli/src/util/client.ts:33]: copied from index.ts

**Comment by codyebberson** [comment]
@tonylee80  I'm conceding defeat on Sonar's warnings about "Prefer using nullish coalescing operator (`??`) instead of a logical or (`||`), as it is a safer operator.".  Please go ahead and use `??`

File: packages/cli/src/util/client.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import { FetchLike, MedplumClient } from '@medplum/core';
+import { FileSystemStorage } from '../storage';
+
+export interface MedplumClientCommandOptions {
```
Comment (codyebberson) [inline packages/cli/src/util/client.ts:4]: Is this interface necessary, or could we just use `MedplumClientOptions`?

Comment (tonylee80) [reply packages/cli/src/util/client.ts:4]: This interface is not necessary. Another idea was to inherir from `MedplumClientOptions`, to include `command` specific global options such as `clientId` and `clientSecret`. `clientId` and `clientSecret` could also be added to `MedplumClientOptions` to be used more broadly by the SDK.

**Comment by tonylee80** [comment]
> @tonylee80 I'm conceding defeat on Sonar's warnings about "Prefer using nullish coalescing operator (`??`) instead of a logical or (`||`), as it is a safer operator.". Please go ahead and use `??`

So much for performance gains 🤣 ... I'll switch toe Sonar's preference.

**Review by codyebberson — APPROVED** [review]

**Pull request merged**